docs: vector sdk for php (#373)
upstash/docs: vector/sdks/php/commands/upsert-vectors.mdx

---
title: Upserting Vectors
---

Every database needs data, and your vector database is no exception.

Our SDK makes it easy to insert or update (upsert) vector data in your indexes and/or namespaces.

## Upsert

With Upstash Vector, you can upsert one or more vectors. For now, let’s focus on upserting a single vector.

> We’ll use the `upsert()` method to instruct the SDK to upsert vectors into an index, as shown below:

Reviewer: ```suggestion
We’ll use the `upsert()` method to instruct the SDK to upsert or update vectors into an index, as shown below:
```

Reviewer: I think we can have an `Update` section after the `Upsert Many` that explains you can update your vectors simply with these operations.

```php
use Upstash\Vector\Index;
use Upstash\Vector\VectorUpsert;

$index = new Index(
  url: "<UPSTASH_VECTOR_REST_URL>",
  token: "<UPSTASH_VECTOR_REST_TOKEN>",
);

$index->upsert(new VectorUpsert(
  id: '1',
  vector: [0.1, 0.2, ...], // "..." represents the dimension size of your vector index.
));
// or within a namespace

> $index->namespace('my-namespace')->upsert(new VectorUpsert(

Reviewer: As mentioned before, maybe we can turn this into a Code Group and have a separate tab for namespace.

Author: Done

  id: '1',
  vector: [0.1, 0.2, ...],
));
```

You can also enhance your index by adding metadata, enabling more efficient filtering in the future.

You can read more about [Metadata](/vector/features/metadata#metadata), [Data](/vector/features/metadata#data) and [Metadata Filtering](/vector/features/filtering) on our docs.

## Upsert Many

Building on the previous section, Upstash Vector also supports upserting multiple vectors at once.

To do this, we’ll use the `upsertMany()` method, which allows you to efficiently insert or update multiple vectors into an index, as shown below:

```php
use Upstash\Vector\Index;
use Upstash\Vector\VectorUpsert;

$index = new Index(
  url: "<UPSTASH_VECTOR_REST_URL>",
  token: "<UPSTASH_VECTOR_REST_TOKEN>",
);

$index->upsertMany([
  new VectorUpsert(
    id: '1',
    vector: [0.1, 0.2, ...],
  ),
  new VectorUpsert(
    id: '2',
    vector: [0.3, 0.1, ...],
  ),
]);
```

> Upserting multiple records simultaneously improves performance by allowing you to batch your imports efficiently.

Reviewer: Can we revisit this sentence? We can use something else instead of `imports`. Maybe we can also provide a better explanation of why batch requests improve the performance, such as reducing network overheads. I would think someone from the core team may provide a better explanation.

<Note>For optimal results, we recommend limiting each batch to no more than 1,000 records at a time.</Note>

## Sparse Indexes

> If your vector index is configured to use sparse indexes, you’ll need to modify your upsert call accordingly.

Reviewer: ```suggestion
If you are using a sparse index, you’ll need to modify your upsert call accordingly.
```

Sparse indexes require a set of indices and their corresponding values, which can be upserted as demonstrated below:

```php
use Upstash\Vector\Index;
use Upstash\Vector\VectorUpsert;
use Upstash\Vector\SparseVector;

$index = new Index(
  url: "<UPSTASH_VECTOR_REST_URL>",
  token: "<UPSTASH_VECTOR_REST_TOKEN>",
);

$index->upsert(new VectorUpsert(
  id: '1',
  sparseVector: new SparseVector(
    indices: [0, 1],
    values: [1.0, 2.0],
  ),
));
```

You can read more about [Sparse Indexes](/vector/features/sparseindexes) on our docs.

## Hybrid Indexes

> If your vector index is configured to use hybrid indexes, you need to provide both sparse vectors and dense vectors.

Reviewer: ```suggestion
If you are using a hybrid index, you need to provide both sparse vectors and dense vectors.
```
